ably/realtime_conn.go

```diff
@@ -18,27 +18,32 @@ var (
 // Conn represents a single connection RealtimeClient instantiates for
 // communication with Ably servers.
 type Conn struct {
-	details       proto.ConnectionDetails
-	id            string
-	serial        int64
-	msgSerial     int64
-	err           error
-	conn          proto.Conn
-	opts          *ClientOptions
-	state         *stateEmitter
-	stateCh       chan State
-	pending       pendingEmitter
-	queue         *msgQueue
-	auth          *Auth
-	onChannelMsg  func(*proto.ProtocolMessage)
-	onStateChange func(State)
+	details      proto.ConnectionDetails
+	id           string
+	serial       int64
+	msgSerial    int64
+	err          error
+	conn         proto.Conn
+	opts         *ClientOptions
+	state        *stateEmitter
+	stateCh      chan State
+	pending      pendingEmitter
+	queue        *msgQueue
+	auth         *Auth
+	onChannelMsg func(*proto.ProtocolMessage)
+	// This callback is called when we get a response from reconnect request. We
+	// move this up because some implementation details for (RTN15c) requires
+	// access to Channels and we dont have it here so we let RealtimeClient do the
+	// work.
+	onReconnectMsg func(*proto.ProtocolMessage)
+	onStateChange  func(State)
 	// reconnecting tracks if we have issued a reconnection request. If we receive any message
 	// with this set to true then its the first message/response after issuing the
 	// reconnection request.
 	reconnecting bool
 }
 
-func newConn(opts *ClientOptions, auth *Auth, onChannelMsg func(*proto.ProtocolMessage), onStateChange func(State)) (*Conn, error) {
+func newConn(opts *ClientOptions, auth *Auth, onChannelMsg, onReconnectMsg func(*proto.ProtocolMessage), onStateChange func(State)) (*Conn, error) {
```

Comment:
Suggestion: when we're at three callbacks, maybe it's time to take a single "delegate" struct-of-callbacks instead?

Reply:
Done at f575c81464759aa2fa48e3221b5845eb8acc3ce6.